### PR TITLE
feat: implement fal_results.json

### PR DESCRIPTION
**Pull request by isidentical**
Implements the `fal_results.json`. It is basically a mirror of the original DBT run results, but contains additional information for `post-hook`s and Python model scripts (including execution time, their name, executed thread, etc.).

### CODE REVIEW

File: src/fal/cli/flow_runner.py [hunk 1]
```diff
@@ -225,47 +231,41 @@ def _unique_ids_to_model_names(id_list: List[str]) -> List[str]:
...
+        results = _get_all_result_content(path)
 
-    all_results = reduce(
-        lambda all, next: all.extend(_get_result_array(next)) or all,
-        result_files,
-        [],
-    )
+        if results.get(SCHEMA_KEY) != "fal":
+            dbt_run_results.append(results)
+        fal_run_results.append(results)
```
Comment (mederka) [inline src/fal/cli/flow_runner.py:244]: Shouldn't there be an else here?

File: src/fal/cli/flow_runner.py [hunk 2]
```diff
@@ -225,47 +231,41 @@ def _unique_ids_to_model_names(id_list: List[str]) -> List[str]:
...
+    # the rest of the run results.
+    result_framework = dbt_run_results[-1]
 
+    for file, results in [
+        (DBT_RUN_RESULTS_FILENAME, dbt_run_results),
+        (FAL_RUN_RESULTS_FILENAME, fal_run_results),
+    ]:
+        combined_results = copy.deepcopy(result_framework)
+        combined_results[RUN_RESULTS_KEY] = [
+            run_info for result in results for run_info in result[RUN_RESULTS_KEY]
```
Comment (mederka) [inline src/fal/cli/flow_runner.py:262]: Maybe we can make this a bit easier to read.

File: src/fal/utils.py [hunk 3]
```diff
@@ -4,6 +4,21 @@
 from typing import List
 
 
+class _IndexFactory:
```
Comment (mederka) [inline src/fal/utils.py:7]: Yay our own iterator!

**Review by mederka — COMMENTED** [review]
Thank you, @isidentical. I just have a couple of comments.

File: src/fal/cli/flow_runner.py (hunk 1 above)
Comment (isidentical) [reply src/fal/cli/flow_runner.py:244]: I think it opens up the question of what 'fal_results.json' is. Does it only contain Fal's own stuff (like post-hooks and Python model script executions) or does it contain both DBT run results and fal's stuff (like a fully fledged thing). WDYT?

CC: @turbo1912 @chamini2

**Review by mederka — APPROVED** [review]

File: src/fal/run_scripts.py [hunk 4]
```diff
@@ -53,15 +85,19 @@ def run_scripts(scripts: List[FalScript], faldbt: FalDbt) -> List[bool]:
         pool: Pool = pool
         try:
             scripts_with_project = map(lambda script: (script, faldbt), scripts)
-            results = pool.starmap(_prepare_exec_script, scripts_with_project)
+            results = pool.starmap(run_script, scripts_with_project)
```
Comment (chamini2) [inline src/fal/run_scripts.py:88]: Eventually we will remove this threading part, right?

**Review by chamini2 — COMMENTED** [review]
Curious about some integration tests or add some examples in the PR comments.

File: src/fal/run_scripts.py (hunk 4 above)
Comment (isidentical) [reply src/fal/run_scripts.py:88]: Yes!

File: src/fal/run_scripts.py [hunk 5]
```diff
@@ -15,22 +19,48 @@
...
+        "thread_id": threading.current_thread().name,
+        "is_post_hook": script._is_post_hook,
+        "execution_time": (finished_at - started_at).total_seconds(),
+        "timing": [
+            {
+                "name": "execute",
+                # DBT's suffix for UTC is Z, but isoformat() uses +00:00. So
+                # we'll manually cast it to the proper format.
+                # https://stackoverflow.com/a/42777551
+                "started_at": started_at.isoformat().replace("+00:00", "Z"),
```
[the remark this answers is not in the document]
Comment (isidentical) [reply src/fal/run_scripts.py:53]: Thanks!!

File: src/fal/cli/flow_runner.py [hunk 6]
```diff
@@ -236,47 +242,41 @@ def _unique_ids_to_model_names(id_list: List[str]) -> List[str]:
...
-    last_run_result = _get_all_result_content(result_files[-1])
-    last_run_result[RUN_RESULTS_KEY] = all_results
-    run_result_file = os.path.join(target_path, RUN_RESULTS_FILE_NAME)
+        # Clear out files as we go.
+        path.unlink()
 
-    # remove all run_results_*.json files for the next fal flow run
-    _remove_all(result_files)
+    if not dbt_run_results:
+        return
```
[the remark this answers is not in the document]
Comment (isidentical) [reply src/fal/cli/flow_runner.py:261]: If we have a pure-fal-script run, no. But python-models (or post-hooks) should always generate a `fal_results.json`.

[the remark this answers is not in the document]
Comment (isidentical) [reply src/fal/cli/flow_runner.py:260]: Created FEA-259 for this.

**Comment by isidentical** [comment]
@chamini2 an example `fal_results.json` file: https://gist.github.com/isidentical/b83eed5c8634e3e18965cd92238c2f61

**Comment by isidentical** [comment]
This heavily touches on the flow runner internals, so I'll wait @chamini2's PRs to be first merged and then open this with a new rebase.

File: src/fal/cli/flow_runner.py [hunk 7]
```diff
@@ -83,47 +84,41 @@ def node_to_script(node: Union[FalFlowNode, None], fal_dbt: FalDbt) -> FalScript
         raise Exception(f"Cannot convert node to script. Node: {node}")
 
 
-def _combine_fal_run_results(target_path):
-    result_files = _get_fal_run_results(target_path)
+def _combine_fal_run_results(target_path: str) -> None:
+    target_path = Path(target_path)
+    dbt_run_results, fal_run_results = [], []
+    for path in target_path.glob("fal_results_*.json"):
+        assert path.is_file()
```
Comment (mederka) [inline src/fal/cli/flow_runner.py:91]: What would it mean for a path to not be a file? Should we catch this more gently?

File: src/fal/cli/flow_runner.py [hunk 8]
```diff
@@ -83,47 +84,41 @@ def node_to_script(node: Union[FalFlowNode, None], fal_dbt: FalDbt) -> FalScript
...
-    if not result_files:
-        return
+        results = _get_all_result_content(path)
 
-    all_results = reduce(
-        lambda all, next: all.extend(_get_result_array(next)) or all,
-        result_files,
-        [],
-    )
+        if results.get(SCHEMA_KEY) != "fal":
```
Comment (mederka) [inline src/fal/cli/flow_runner.py:95]: Nit: Maybe we put "fal" in a variable?

File: src/fal/cli/flow_runner.py [hunk 9]
```diff
@@ -83,47 +84,41 @@ def node_to_script(node: Union[FalFlowNode, None], fal_dbt: FalDbt) -> FalScript
...
+        combined_results = copy.deepcopy(result_framework)
+        combined_results[RUN_RESULTS_KEY] = []
+        for result in results:
+            combined_results[RUN_RESULTS_KEY].extend(result[RUN_RESULTS_KEY])
 
-def _remove_all(files: List[str]):
-    for file in files:
-        os.remove(file) if os.path.exists(file) else None
+        with open(target_path / file, "w") as stream:
+            json.dump(combined_results, stream)
```
Comment (mederka) [inline src/fal/cli/flow_runner.py:119]: Nit: this can be more readable

File: src/fal/cli/flow_runner.py (hunk 8 above)
Comment (chamini2) [reply src/fal/cli/flow_runner.py:95]: this is checking the `"schema"` property in the `fal_results_X.json` file. If it DOES NOT have `"schema": "fal", we process it as a `dbt` run_results artifact.

Comment (chamini2) [reply src/fal/cli/flow_runner.py:95]: which gets me thinking, @isidentical , this is built for `"dbt_schema_version": "https://schemas.getdbt.com/dbt/run-results/v4.json"`. Maybe we should check that it is that scema version?

**Review by chamini2 — APPROVED** [review]
👏

File: src/fal/cli/flow_runner.py (hunk 7 above)
Comment (isidentical) [reply src/fal/cli/flow_runner.py:91]: > What would it mean for a path to not be a file? Should we catch this more gently?

I don't think so, this is just a sanity check (you can create a directory named `fal_results_1.json` but I don't think it should be a case we should try to handle).

File: src/fal/cli/flow_runner.py (hunk 8 above)
Comment (isidentical) [reply src/fal/cli/flow_runner.py:95]: Makes sense!

File: src/fal/cli/flow_runner.py (hunk 9 above)
Comment (isidentical) [reply src/fal/cli/flow_runner.py:119]: Not sure what you mean here @mederka?

**Pull request merged**